### PR TITLE
Add Hi Goodtimes statistical filter 1 check

### LINKED ISSUE #2724: Hi Goodtimes - statistical filter 1
### Algorithm Description:

Implement the statistical filter 1 from the Hi algorithm document.

### PR DESCRIPTION
**Pull request by tmplummer**
# Change Summary

## Summary
                                                                                                                                                                                                                                                                
  Implements Statistical Filter 1 for Hi Goodtimes processing, which detects isotropic count rate increases per Algorithm Document Section 2.3.2.3.

## New Functions

  - _compute_qualified_counts_per_sweep(): Computes qualified calibration product counts per 8-spin interval using vectorized np.add.at() with the (esa_sweep, esa_step) multi-index
  - _build_per_sweep_datasets(): Builds per-sweep datasets with qualified counts for multiple Pointings
  - _compute_median_and_sigma_per_esa(): Computes median and sigma per ESA step using xr.concat and xr.median to handle datasets with potentially different ESA step coordinates
  - _identify_cull_pattern(): Identifies 2D cull patterns using scipy.ndimage.convolve1d for efficient pattern detection
  - mark_statistical_filter_1(): Main entry point that applies the three-pass filter algorithm

## Algorithm

  The filter applies three passes using 2D convolution:
  1. Consecutive runs: Marks intervals where counts exceed median + 1.8σ for 3+ consecutive intervals AND the adjacent ESA step also exceeds threshold at the same time
  2. Isolated intervals: Marks good intervals sandwiched between two bad intervals (orphan detection)
  3. Extreme outliers: Marks remaining intervals exceeding median + 5σ

## Implementation Highlights

  - Uses idiomatic xarray operations (xr.concat, xr.median, broadcasting) instead of loops
  - Leverages scipy.ndimage.convolve1d for efficient pattern matching along both time (esa_sweep) and energy (esa_step) dimensions
  - Handles datasets with different ESA step coordinates by using xr.concat which aligns and fills missing values with NaN
  - Returns xr.DataArray objects with proper coordinates for seamless integration

## Test Coverage

  Adds comprehensive tests for all new functions including edge cases for empty data, NaN handling, and pattern detection verification.

## File changes
- imap_processing/hi/hi_goodtimes.py
- imap_processing/tests/hi/test_hi_goodtimes.py


Closes: #2724

### CODE REVIEW

File: imap_processing/hi/hi_goodtimes.py [hunk 1]
```diff
@@ -1078,3 +1079,410 @@ def mark_statistical_filter_0(
...
+    # Add qualified_count - aligns with (esa_sweep, esa_step) coordinates
+    ds_reshaped["qualified_count"] = xr.DataArray(
+        counts_2d,
+        dims=["esa_sweep", "esa_step"],
+        coords={
+            "esa_sweep": np.arange(n_sweeps),
+            "esa_step": np.arange(n_esa_steps),
+        },
+    )
+
```
[the remark this answers is not in the document]
Comment (tmplummer) [reply imap_processing/hi/hi_goodtimes.py:1145]: Good catch.

File: imap_processing/hi/hi_goodtimes.py [hunk 2]
```diff
@@ -1078,3 +1079,406 @@ def mark_statistical_filter_0(
...
+    is_qualified = np.isin(coincidence_type, list(qualified_coincidence_types))
+
+    # Map qualified events to their packet's (esa_sweep, esa_step)
+    qualified_packet_idx = ccsds_index[is_qualified]
+    qualified_sweep = esa_sweep[qualified_packet_idx]
+    qualified_step = esa_step[qualified_packet_idx]
+
+    # Count qualified events per (esa_sweep, esa_step) using 2D array
+    n_sweeps = int(esa_sweep.max()) + 1
+    n_esa_steps = int(esa_step.max()) + 1
```
Comment (laspsandoval) [inline imap_processing/hi/hi_goodtimes.py]: Does esa_sweep start at zero each time?

File: imap_processing/hi/hi_goodtimes.py [hunk 3]
```diff
@@ -1078,3 +1079,406 @@ def mark_statistical_filter_0(
...
+    counts_2d = np.zeros((n_sweeps, n_esa_steps), dtype=np.float64)
```
Comment (laspsandoval) [inline imap_processing/hi/hi_goodtimes.py]: `# For every pair of indices (sweep_i, step_i), add 1 to that cell.`

File: imap_processing/hi/hi_goodtimes.py [hunk 4]
```diff
@@ -1078,3 +1079,406 @@ def mark_statistical_filter_0(
...
+def mark_statistical_filter_1(
+    goodtimes_ds: xr.Dataset,
+    l1b_de_datasets: list[xr.Dataset],
+    current_index: int,
+    qualified_coincidence_types: set[int],
+    consecutive_threshold_sigma: float = 1.8,
+    extreme_threshold_sigma: float = 5.0,
+    min_consecutive_intervals: int = 3,
+    cull_code: int = CullCode.LOOSE,
+    min_pointings: int = 4,
```
Comment (laspsandoval) [inline imap_processing/hi/hi_goodtimes.py]: Curious why you chose 4?

File: imap_processing/hi/hi_goodtimes.py [hunk 5]
```diff
@@ -1078,3 +1079,406 @@ def mark_statistical_filter_0(
...
+        return empty, empty.astype(int)
+
+    # Concatenate datasets along esa_sweep dimension using xarray
+    # This handles different esa_step coordinates by aligning and filling with NaN
+    combined = xr.concat(
+        [ds["qualified_count"] for ds in per_sweep_datasets.values()],
+        dim="esa_sweep",
+    )
+
+    # Compute median along esa_sweep dimension using xarray
```
Comment (laspsandoval) [inline imap_processing/hi/hi_goodtimes.py]: Clarify:
`# Median across all sweeps and pointings.`

File: imap_processing/hi/hi_goodtimes.py [hunk 6]
```diff
@@ -1078,3 +1079,406 @@ def mark_statistical_filter_0(
...
+            f"sigma={sigma_per_esa.sel(esa_step=esa).values}"
+        )
+
+    return median_per_esa, sigma_per_esa
+
+
+def _identify_cull_pattern(
+    current_counts: xr.DataArray,
+    median_per_esa: xr.DataArray,
+    sigma_per_esa: xr.DataArray,
```
Comment (laspsandoval) [inline imap_processing/hi/hi_goodtimes.py:1260]: Why these defaults?

File: imap_processing/hi/hi_goodtimes.py [hunk 7]
```diff
@@ -1078,3 +1079,406 @@ def mark_statistical_filter_0(
...
+
+    # === Pass 1: Find consecutive runs with ESA neighbor confirmation ===
+    # Use convolution to find runs of min_consecutive in time (axis=0 = esa_sweep)
+    time_kernel = np.ones(min_consecutive)
+    consecutive_sum = convolve1d(exceeds_arr, time_kernel, axis=0, mode="constant")
+
+    # Dilate the consecutive detection to mark all positions in runs
+    # convolve1d centers the kernel, so we dilate to capture run edges
+    run_kernel = np.ones(min_consecutive)
+    run_positions = convolve1d(
```
Comment (laspsandoval) [inline imap_processing/hi/hi_goodtimes.py:1320]: Nice way of doing this.

**Review by laspsandoval — APPROVED** [review]
I went through very carefully and everything looks good. Just some comments/questions about constant values. I'm approving since there isn't anything significant  that needs to be changed. Nice work!

File: imap_processing/hi/hi_goodtimes.py (hunk 2 above)
Comment (tmplummer) [reply imap_processing/hi/hi_goodtimes.py]: Thanks for this comment. It made me realize that I have a bug. The esa_sweep dimension is somewhat arbitrary. It just counts how many times the ESA steps through the full range of energy steps (nominally 9-steps). But the bug is that the definition of which energy esa_step N corresponds to can change from one pointing to the next. So, I should be using esa_energy_step here in order to make sure that the correct energy steps get aligned.

File: imap_processing/hi/hi_goodtimes.py (hunk 6 above)
Comment (tmplummer) [reply imap_processing/hi/hi_goodtimes.py:1260]: Good point. These are from the algorithm document, but should be added to the Hi Constants.
